pallet-ismp/evm/src/handler.rs

```diff
@@ -0,0 +1,213 @@
+                    source: get.source.to_string().as_bytes().to_vec(),
+                    dest: get.dest.to_string().as_bytes().to_vec(),
+                    nonce: u64_to_u256(get.nonce)?,
+                    height: u64_to_u256(get.height)?,
+                    timeoutTimestamp: u64_to_u256(get.timeout_timestamp)?,
+                    from: get.from,
+                    keys: get.keys,
+                };
+                (OnGetTimeoutCall { request }.encode(), gas_limit)
+            }
+        };
+        execute_call::<T>(target_contract, call_data, gas_limit)
+    }
+}
+
+/// Parse contract id from raw bytes
+pub fn parse_contract_id(bytes: &[u8]) -> Result<H160, Error> {
+    let module_id =
+        ModuleId::from_bytes(bytes).map_err(|e| Error::ImplementationSpecific(e.to_string()))?;
+    match module_id {
+        ModuleId::Evm(id) => Ok(id),
+        _ => Err(Error::ImplementationSpecific("Expected Evm contract id".to_string())),
+    }
+}
+
+/// Convert u64 to U256
+pub fn u64_to_u256(value: u64) -> Result<U256, Error> {
+    U256::try_from(value)
+        .map_err(|_| Error::ImplementationSpecific("Failed to convert u64 to u256".to_string()))
+}
+
+/// Call execute call data
+fn execute_call<T: pallet_ismp::Config + pallet_evm::Config>(
+    target: H160,
+    call_data: Vec<u8>,
+    gas_limit: u64,
+) -> Result<(), Error> {
+    match <<T as pallet_evm::Config>::Runner as pallet_evm::Runner<T>>::call(
+        EVM_HOST_ADDRESS,
+        target,
+        call_data,
+        Default::default(),
+        gas_limit,
+        None,
+        None,
+        None,
+        Default::default(),
+        true,
+        true,
+        None,
+        None,
+        <T as pallet_evm::Config>::config(),
+    ) {
+        Ok(info) => {
+            let mut total_weight_used = WeightConsumed::<T>::get();
+            let weight_limit = T::GasWeightMapping::gas_to_weight(gas_limit, true);
+            let weight_used =
+                T::GasWeightMapping::gas_to_weight(info.used_gas.standard.low_u64(), true);
+            total_weight_used.weight_used = total_weight_used.weight_used + weight_used;
+            total_weight_used.weight_limit = total_weight_used.weight_limit + weight_limit;
+            WeightConsumed::<T>::put(total_weight_used);
+            Ok(())
+        }
+        Err(error) => {
+            let mut total_weight_used = WeightConsumed::<T>::get();
+            let weight_limit = T::GasWeightMapping::gas_to_weight(gas_limit, true);
+            total_weight_used.weight_used = total_weight_used.weight_used + error.weight;
+            total_weight_used.weight_limit = total_weight_used.weight_limit + weight_limit;
+            WeightConsumed::<T>::put(total_weight_used);
+            // We still return ok so we can compensate for used gas only
+            Err(Error::ImplementationSpecific(
+                "Contract encountered error while executing".to_string(),
+            ))
+        }
```

Comment:
prefer if we dedupe this match case.
```suggestion
let weight_used = match result {
    Ok(info) => T::GasWeightMapping::gas_to_weight(info.used_gas.standard.low_u64(), true),
    Err(error) => error.weight,
}

let mut total_weight_used = WeightConsumed::<T>::get();
let weight_limit = T::GasWeightMapping::gas_to_weight(gas_limit, true);
total_weight_used.weight_used = total_weight_used.weight_used + weight_used;
total_weight_used.weight_limit = total_weight_used.weight_limit + weight_limit;
WeightConsumed::<T>::put(total_weight_used);
```